### PR TITLE
Create optional animation and select my show frame #44

### PR DESCRIPTION
**Pull request by gpaixao**
Before when created sprite instance you just set `loop: false || true`, but if I wan't animate my sprite ?

### Create animation in loop
```js
origami("#mcanvas")
.background("blue")
.sprite(0, 0, {
  src: 'sprite.png',
  frames: 1,
  speed: 100,
  animation: true,
  loop: true
 })
 .load(function (octx) { octx.draw(); });
```

### Without animation and select show frame
```js
origami("#mcanvas")
.background("blue")
.sprite(0, 0, {
  src: 'sprite.png',
  frames: 2,
  frame: 2,
  animation: false
 })
 .load(function (octx) { octx.draw(); });
```

### CODE REVIEW

**Comment by raphamorim** [comment]
Hi @gpaixao, thanks for your time and contribution. I highlighted some notes about your PR:

- `currentFrame` sound better than `frame`
- check my notes on the code

File: dist/origami.js [hunk 1]
```diff
@@ -716,55 +715,55 @@ Origami.border = function() {
...
+};
 
 function SpriteShape(params) {
   var properties = params.properties,
     dw = params.width / properties.frames;
 
   drawSprite.call(this, {
     image: params.image,
-    posX: 0,
+    posX: (properties.frame > 0 && properties.frame <= properties.frames ? (dw *(properties.frame -1)) : 0),
```
Comment (raphamorim) [inline dist/origami.js]: Put this logic on variable instead put in Object declaration

File: src/shapes/sprite.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ function SpriteShape(params) {
 
   drawSprite.call(this, {
     image: params.image,
-    posX: 0,
+    posX: (properties.frame > 0 && properties.frame <= properties.frames ? (dw *(properties.frame -1)) : 0),
```
Comment (raphamorim) [inline src/shapes/sprite.js]: Put this logic on variable instead put in Object declaration

File: src/shapes/sprite.js [hunk 3]
```diff
@@ -14,16 +14,18 @@ function SpriteShape(params) {
     dx: params.x,
     dy: params.y,
     speed: properties.speed,
-    animation: null
+    update: null,
+    animation: properties.animation === true ? true : false,
```
Comment (raphamorim) [inline src/shapes/sprite.js]: use `Boolean(properties.animation)`

File: src/shapes/sprite.js (hunk 2 above)
Comment (gpaixao) [inline src/shapes/sprite.js]: HI, @raphamorim . I think this. Great!

File: src/shapes/sprite.js (hunk 3 above)
Comment (raphamorim) [inline src/shapes/sprite.js]: I think `animation: properties.animation,` will work

File: src/shapes/sprite.js [hunk 4]
```diff
@@ -14,16 +14,18 @@ function SpriteShape(params) {
...
+    loop: properties.loop === true ? true : false
```
Comment (raphamorim) [inline src/shapes/sprite.js]: Same here

File: src/shapes/sprite.js (hunk 3 above)
Comment (gpaixao) [inline src/shapes/sprite.js]: But if `properties.animation` is equal `string` ? It return `true`, but isn't `true`.

File: src/shapes/sprite.js (hunk 4 above)
Comment (raphamorim) [inline src/shapes/sprite.js]: Doesn't matter if it's a string, run this code in your console

```js
var properties = {};
properties.animation = "some-random-string";
if (properties.animation) {
   console.log("properties.animation is true");
}
```

File: src/shapes/sprite.js (hunk 3 above)
Comment (raphamorim) [inline src/shapes/sprite.js]: Doesn't matter if it's a string, run this code in your console

```js
var properties = {};
properties.animation = "some-random-string";
if (properties.animation) {
   console.log("properties.animation is true");
}
```

Comment (raphamorim) [inline src/shapes/sprite.js]: In cases like this, doesn't matter if is a **real** boolean, if use set/or/create animation params will be true

Comment (raphamorim) [inline src/shapes/sprite.js]: But if you REALLY want to check if it's true
You can rewrite to
```js
animation: Boolean(properties.animation)
```

Comment (gpaixao) [inline src/shapes/sprite.js]: But not is better normalize input for **bool**? If give a **string** like param, my variable will is `true` but it's not a **bool**.

```js
var properties = {};
properties.animation = "some-random-string"; // typeof string, but true in condition.
if (properties.animation) {
   console.log("properties.animation is true");
}
```

Comment (raphamorim) [inline src/shapes/sprite.js]: There's not problem with that, we have to work with language characteristics. JS has weak types and we have to accept it

```js
"" === true // true
```

Comment (gpaixao) [inline src/shapes/sprite.js]: So, if the client set:

```js
{
  animation: "teste"
}
```

Animation work, because **"teste"** is equal **true**.

Comment (raphamorim) [inline src/shapes/sprite.js]: yes, that's it :)

Comment (gpaixao) [inline src/shapes/sprite.js]: would be right?

Comment (raphamorim) [inline src/shapes/sprite.js]: "teste" in js is `true`, right?
Since we write test for all cases (Fall and success case, there's no problem with that)

Comment (gpaixao) [inline src/shapes/sprite.js]: Great!!! 😄

Comment (raphamorim) [inline src/shapes/sprite.js]: Nice :octocat: :)

Comment (raphamorim) [inline src/shapes/sprite.js]: hehe I know what you fell, unfortunately hehe

File: src/shapes/sprite.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ function SpriteShape(params) {
 
   drawSprite.call(this, {
     image: params.image,
-    posX: (properties.frame > 0 && properties.frame <= properties.frames ? (dw *(properties.frame -1)) : 0),
+    posX: (properties.currentFrame > 0 && properties.currentFrame <= properties.frames ? (dw *(properties.currentFrame -1)) : 0),
```
Comment (raphamorim) [inline src/shapes/sprite.js]: Only missing put this logic in a variable

**Comment by raphamorim** [comment]
Hi @gpaixao can you write tests for this change?

**Comment by gpaixao** [comment]
Sorry, I'm commited fail test. I'm change here !

File: test/test.html [hunk 6]
```diff
@@ -32,7 +32,7 @@
         <!-- <script src="./test-draw-shape-css.js"></script> -->
         <script src="./test-smart-coordinates.js"></script>
         <script src="./test-components.js"></script>
-        <script src="./test-plugins.js"></script>
+        <!--<script src="./test-plugins.js"></script>-->
```
Comment (gpaixao) [inline test/test.html]: I comment because test not found running local.

Comment (gpaixao) [inline test/test.html]: I change here.

**Comment by raphamorim** [comment]
Amazing I'll review it tomorrow.
Can you send a PR to [origamijs.com](https://github.com/raphamorim/origamijs.com) updating sprite docs?

**Comment by raphamorim** [comment]
Hi @gpaixao,
I put your commits in another branch and fixed all tests (they always are passing, cause 100ms for interval isn't enough to render in canvas and compare images)

Also I've redesign the Sprite API too. I want know what you think about the new Sprite API.

Thanks for your contribution/time 👍

**Comment by gpaixao** [comment]
Hi, @raphamorim .

In first moment, sorry. I never tested data from canvas. Commited tests failured is very bad. Sorry.

And about redesign of Sprite API I can't think about anything better for API now.

Maybe anything that think is include method for to facilitate create games, like `jump()`, `down()`, `left()` and ...

Maybe this context isn't the best for it, but I miss this methods when used origami.js for create a canvas game.

Because usually its used for create animations and interations like a game.

**Comment by raphamorim** [comment]
No problem with that,

About animation API, I want to Origami 0.6v have a decent animation API based on animation principles: https://github.com/raphamorim/origami.js/issues/37

Take a look, if possible give more ideas :)

**Comment by raphamorim** [comment]
Btw: I've created a little game using origamijs and missed a animation api too

**Pull request merged**